components/raftstore/src/store/peer.rs

```diff
@@ -1855,16 +1869,132 @@ where
+        let leader_store_id = self.peer.get_store_id();
+
+        if self.raft_group.raft.follower_replication() && self.is_leader() && let Some(leader_zone) = ctx.zone_info.get(&leader_store_id) {
+            // Group MsgAppend by AZ.
+            let mut msg_append_group: HashMap<String, Vec<usize>> = HashMap::default();
+            // Record message that should be discarded after merge_msg_append.
+            let mut skip = vec![false; msgs.len()];
+            // Whether or not start group broadcast at this ready.
+            // Actually if msgs does not contain complete messages sent by send_bcast, 
+            // it means that replication is not in a good state.
+            let mut should_forward = true;
+
+            // Filter MsgAppend in msgs.
+            // Follower replication is enabled and msg is MsgAppend.
+            let mut pos: usize = 0;
+            while pos < msgs.len() {
+                let msg = &msgs[pos];
+                if msg.get_msg_type() == MessageType::MsgAppend
+                    && self.raft_group.raft.is_replicating(msg.get_to())
+                {
+                    if let Some(to_peer) = self.get_peer_from_cache(msg.get_to()) {
+                        let to_peer_store_id = to_peer.get_store_id();
+                        // Leader and to_peer is not in the same zone.
+                        if let Some(to_peer_zone) = ctx.zone_info.get(&to_peer_store_id) && to_peer_zone != leader_zone
+                        {
+                            // If batch_append is disabled, there may be several MsgAppend to a same peer in msgs.
+                            // The agent should not forward more than one MsgAppend to a single peer.
+                            if let Some(v) = msg_append_group.get_mut(to_peer_zone) {
+                                for idx in v.iter() {
+                                    if msgs[*idx].to == msgs[pos].to {
+                                        should_forward = false;
+                                        break;
+                                    }
+                                }
+                                v.push(pos);
+                            } else {
+                                msg_append_group.insert(to_peer_zone.clone(), vec![pos]);
+                            }
+                        }
+                    }
+                }
+                pos += 1;
+            }
+            if should_forward {
+                for (_, group) in msg_append_group.iter() {
+                    self.merge_msg_append(group, &mut msgs, &mut skip);
+                }
+            }
+
+            // Filter unnecessary MsgAppend and build raft messages.
+            for (pos, msg) in msgs.into_iter().enumerate() {
+                if !skip[pos] && let Some(m) = self.build_raft_message(msg, ctx.self_disk_usage) {
+                    raft_msgs.push(m);
+                }
+            }
+        } else {
+            for msg in msgs {
```

Comment:
minor: maybe it's better to put else to if as it's more common code path. Just for better readability.